panther_analysis_tool/main.py

```diff
@@ -888,38 +926,97 @@ def run_tests(
+    analysis_funcs: Dict[str, Any],
+) -> Dict[str, Any]:
+    test_run_result: Dict[str, str] = defaultdict(lambda: "PASS")
+    for func in RESERVED_FUNCTIONS:
+        if analysis_funcs.get(func):
+            try:
+                if mock_methods:
+                    with patch.multiple(analysis_funcs["module"], **mock_methods):
+                        func_out = analysis_funcs[func](test_case)
+                        if not func_out:
+                            test_run_result[func] = "FAIL"
+                else:
+                    func_out = analysis_funcs[func](test_case)
+                    if not func_out:
+                        test_run_result[func] = "FAIL"
+            except Exception as err:  # pylint: disable=broad-except
+                func_out = err
+                test_run_result[func] = "FAIL"
+
+            func_out_valid_type = True
+            # severity value validation
+            valid_output, func_out = validate_outputs(func, func_out)
+            if not valid_output or isinstance(func_out, (AttributeError, AssertionError)):
+                func_out_valid_type = False
+                test_run_result[func] = "FAIL"
+
+            print(
+                f"\t\t[{test_run_result[func]}] "
+                f"{'' if func_out_valid_type else '[INVALID TYPE] '}"
+                f"[{func}] {func_out}"
+            )
+    return test_run_result
+
+
+def validate_outputs(function_name: str, function_output: Any) -> (bool, Any):
+    valid = True
+    out = function_output
+    if function_name == "severity":
+        if not isinstance(function_output, str):
+            valid = False
+        if str(function_output).upper() not in VALID_SEVERITIES:
+            valid = False
+            out = AssertionError(
+                f'Expected [{function_name}] to be any of the following: '
+                f'{str(VALID_SEVERITIES)}, got [{str(function_output)}] instead.'
+            )
+    elif function_name == "destinations":
+        if isinstance(function_output, list):
+            if not all(isinstance(x, str) for x in function_output):
+                list_types = set()
+                for each_item in function_output:
+                    list_types.add(type(each_item))
+                valid = False
+                out = AttributeError(
+                    f'Expected [{function_name}] to return a list of strings, '
+                    f'got types [{list_types}] instead'
                 )
-
-    if minimum_tests > 1 and not (
-        [x for x in analysis["Tests"] if x["ExpectedResult"]]
-        and [x for x in analysis["Tests"] if not x["ExpectedResult"]]
-    ):
-        failed_tests[analysis.get("PolicyID") or analysis["RuleID"]].append(
-            "Insufficient test coverage: expected at least one positive and one negative test"
+        else:
+            valid = False
+    else:
+        valid = isinstance(function_output, str)
+
+    if not valid and out == function_output:
```

Comment:
nit: could you rename or add a comment here (or at declaration) on what `out` represents / what this comparison is doing? It was a little hard to follow what is actually being compared here.

Reply:
Yeah this function seems to be doing the correct thing, I also had to re-read a few times to follow along though. I'd suggest either some more comments or maybe some renaming to make it a bit more clear what exactly is going on.